test/drbgtest.c

```diff
@@ -864,13 +865,79 @@ static int test_multi_thread(void)
 }
 #endif
 
+/*
+ * This function is only returns the entropy already added with RAND_add(),
+ * and does not get entropy from the OS.
+ */
+static size_t rand_add_entropy(RAND_DRBG *drbg,
+                               unsigned char **pout,
+                               int entropy, size_t min_len, size_t max_len,
+                               int prediction_resistance)
+{
+    size_t ret = 0;
+
+    if (drbg->pool == NULL)
+        return 0;
+
+    if (drbg->pool->entropy < (size_t)entropy || drbg->pool->len < min_len
+        || drbg->pool->len > max_len)
+        return 0;
+
+    *pout = malloc(drbg->pool->len);
+    if (*pout == NULL)
+        return 0;
+
+    memcpy(*pout, drbg->pool->buffer, drbg->pool->len);
+    ret = drbg->pool->len;
+    OPENSSL_secure_clear_free(drbg->pool->buffer, drbg->pool->max_len);
+    OPENSSL_free(drbg->pool);
+    drbg->pool = NULL;
+
+    return ret;
+}
+
+static void rand_add_clean_entropy(RAND_DRBG *drbg, unsigned char *out, size_t outlen)
```

Comment:
The callbacks `rand_add_entropy()` and `rand_add_clean_entropy()` are named inconsistently (the callback hooks are named `get_entropy()` and `cleanup_entropy()`, respectively). In particular, the name `rand_add_clean_entropy()` suggests that the function is adding "clean entropy".

You could change the names to `rand_add_get_entropy()` and `rand_add_cleanup_entropy()`, but that sounds a little awkward, too. How do you like one of the following two suggestions?

- `get_pool_entropy()` and `cleanup_pool_entropy()`
- `get_added_entropy()` and `cleanup_added_entropy()`